bottlecap/src/lifecycle/invocation/span_inferrer.rs

```diff
@@ -0,0 +1,109 @@
+const FUNCTION_TRIGGER_EVENT_SOURCE_ARN_TAG: &str = "function_trigger.event_source_arn";
+
+pub struct SpanInferrer {
+    inferred_span: Option<Span>,
+    is_async_span: bool,
+}
+
+impl Default for SpanInferrer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SpanInferrer {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            inferred_span: None,
+            is_async_span: false,
+        }
+    }
+
+    /// Given a byte payload, try to deserialize it into a `serde_json::Value`
+    /// and try matching it to a `Trigger` implementation, which will create
+    /// an inferred span and set it to `self.inferred_span`
+    ///
+    pub fn infer_span(&mut self, payload: &[u8], aws_config: &AwsConfig) {
+        self.inferred_span = None;
+        if let Ok(payload_value) = serde_json::from_slice::<Value>(payload) {
+            if APIGatewayHttpEvent::is_match(&payload_value) {
+                if let Some(t) = APIGatewayHttpEvent::new(payload_value) {
+                    let mut span = Span {
+                        span_id: Self::generate_span_id(),
+                        ..Default::default()
+                    };
+
+                    t.enrich_span(&mut span);
+                    span.meta.extend([
+                        (
+                            FUNCTION_TRIGGER_EVENT_SOURCE_TAG.to_string(),
+                            "api_gateway".to_string(),
+                        ),
+                        (
+                            FUNCTION_TRIGGER_EVENT_SOURCE_ARN_TAG.to_string(),
+                            t.get_arn(&aws_config.region),
+                        ),
+                    ]);
+
+                    self.is_async_span = t.is_async();
+                    self.inferred_span = Some(span);
+                }
+            } else {
+                debug!("Unable to infer span from payload");
+            }
+        } else {
+            debug!("Unable to serialize payload");
+        }
+    }
+
+    /// If a `self.inferred_span` exist, set the `parent_id` to
+    /// the span.
+    ///
+    pub fn set_parent_id(&mut self, parent_id: u64) {
+        if let Some(s) = &mut self.inferred_span {
+            s.parent_id = parent_id;
+        }
+    }
+
+    pub fn complete_inferred_span(&mut self, invocation_span: &Span) {
+        if let Some(s) = &mut self.inferred_span {
+            if self.is_async_span {
+                if s.duration != 0 {
+                    let duration = invocation_span.start - s.start;
+                    s.duration = duration;
+                }
+            } else {
+                let duration = (invocation_span.start + invocation_span.duration) - s.start;
+                s.duration = duration;
+            }
+
+            s.trace_id = invocation_span.trace_id;
+        }
+    }
+
+    fn generate_span_id() -> u64 {
+        // todo: secure random id with OsRng for SnapStart
+        let mut rng = rand::thread_rng();
```

Comment:
we should do this before merge, I can take a look if you want

Reply:
Ok, should be a simple fix, will look into it tomorrow – what do you mean by `tags like cold start`?

Reply:
You've got a comment in the code about adding tags for cold starts/proactive inits on the span